File: zinnia/compile/optim_pass/memory_trace_injection.py
import copy

from zinnia.compile.builder.builder_impl import IRBuilderImpl
from zinnia.compile.ir.ir_graph import IRGraph
from zinnia.compile.optim_pass.abstract_pass import AbstractIRPass
from zinnia.ir_def.defs.ir_memory_trace_emit import MemoryTraceEmitIR
from zinnia.ir_def.defs.ir_memory_trace_seal import MemoryTraceSealIR
from zinnia.ir_def.defs.ir_read_memory import ReadMemoryIR
from zinnia.ir_def.defs.ir_write_memory import WriteMemoryIR


class MemoryTraceInjectionIRPass(AbstractIRPass):
    def __init__(self):
        super().__init__()

    def exec(self, ir_graph: IRGraph) -> IRGraph:
        ir_graph = copy.copy(ir_graph)
        topological_order = ir_graph.get_topological_order(False)
        in_links, _ = ir_graph.get_io_links()
        values_lookup = {}
        ir_builder = IRBuilderImpl()
        has_memory_access = False

        for stmt in topological_order:
            ir_args = [values_lookup[arg] for arg in in_links[stmt.stmt_id]]
            new_val = ir_builder.create_ir(stmt.ir_instance, ir_args, None)
            values_lookup[stmt.stmt_id] = new_val

            if isinstance(stmt.ir_instance, WriteMemoryIR):
                has_memory_access = True
                ir_builder.create_ir(
                    MemoryTraceEmitIR(stmt.ir_instance.segment_id, is_write=True),
                    [ir_args[0], ir_args[1]],
                    None,
                )
            elif isinstance(stmt.ir_instance, ReadMemoryIR):
                has_memory_access = True
                ir_builder.create_ir(
                    MemoryTraceEmitIR(stmt.ir_instance.segment_id, is_write=False),
                    [ir_args[0], new_val],
                    None,
                )

        if has_memory_access:
            ir_builder.create_ir(MemoryTraceSealIR(), [], None)

        return ir_builder.export_ir_graph()

File: src/optim/memory_trace_injection.rs
use std::collections::HashMap;

use crate::builder::IRBuilder;
use crate::ir::IRGraph;
use crate::ir_defs::IR;
use crate::types::{StmtId, Value};

use super::IRPass;

pub struct MemoryTraceInjection;

impl IRPass for MemoryTraceInjection {
    fn exec(&self, ir_graph: IRGraph) -> IRGraph {
        let mut builder = IRBuilder::new();
        let mut value_lookup: HashMap<StmtId, Value> = HashMap::new();
        let mut has_memory_access = false;

        for stmt in ir_graph.get_topological_order(false) {
            let ir_args: Vec<Value> = stmt
                .arguments
                .iter()
                .map(|&arg| value_lookup[&arg].clone())
                .collect();

            let new_val = builder.create_ir(&stmt.ir, &ir_args);
            value_lookup.insert(stmt.stmt_id, new_val.clone());

            match &stmt.ir {
                IR::WriteMemory { segment_id } => {
                    has_memory_access = true;
                    builder.create_ir(
                        &IR::MemoryTraceEmit {
                            segment_id: *segment_id,
                            is_write: true,
                        },
                        &[ir_args[0].clone(), ir_args[1].clone()],
                    );
                }
                IR::ReadMemory { segment_id } => {
                    has_memory_access = true;
                    builder.create_ir(
                        &IR::MemoryTraceEmit {
                            segment_id: *segment_id,
                            is_write: false,
                        },
                        &[ir_args[0].clone(), new_val],
                    );
                }
                _ => {}
            }
        }

        if has_memory_access {
            builder.create_ir(&IR::MemoryTraceSeal, &[]);
        }

        builder.export_ir_graph()
    }
}
